QuestionAnswering/QA_SEQ2SEQ/training/core/sparql/graph_pattern.py

```diff
@@ -63,6 +111,67 @@ def __str__(self):
             ),
         )
 
+    @classmethod
+    def extract(cls, sparql_fragment: str):
+        subj, sparql_fragment = sparql_fragment.split(maxsplit=1)
+
+        tails = []
+        while True:
+            sparql_fragment = sparql_fragment.strip()
+            predicate, sparql_fragment = sparql_fragment.split(maxsplit=1)
+
+            sparql_fragment = sparql_fragment.strip()
+            if sparql_fragment.startswith('"'):
+                ptr = 1
+                while ptr < len(sparql_fragment) and sparql_fragment[ptr] != '"':
+                    ptr += 1
+                obj = sparql_fragment[: ptr + 1]
+                sparql_fragment = sparql_fragment[ptr + 1 :].strip()
+            elif sparql_fragment.startswith('['):
+                stack = ['[']
+                ptr = 1
+                obj = None
+                while ptr < len(sparql_fragment):
+                    if sparql_fragment[ptr] in "[":
+                        stack.append('[')
+                    elif sparql_fragment[ptr] == "]":
+                        assert stack[-1] == '[', "Unexpected close square bracket: " + sparql_fragment
+                        if len(stack) == 1:
+                            obj = sparql_fragment[:ptr + 1]
+                            sparql_fragment = sparql_fragment[ptr + 1 :].strip()
+                            break
+                        else:
+                            stack.pop()
+                    elif sparql_fragment[ptr] == '"':
+                        ptr += 1
+                        while ptr < len(sparql_fragment) and sparql_fragment[ptr] != '"':
+                            ptr += 1
+                    else:
+                        pass
+                    ptr += 1
+                assert obj is not None
+            else:
+                splits = sparql_fragment.split(maxsplit=1)
+                if len(splits) == 2:
+                    obj, sparql_fragment = splits
+                else:
+                    (obj,) = splits
+                    sparql_fragment = ""
+
+                if obj.endswith(".") or obj.endswith(";"):
+                    punctuation = obj[-1]
+                    sparql_fragment = punctuation + sparql_fragment
+                    obj = obj[:-1]
+
+            tails.append((predicate, obj))
+
+            assert sparql_fragment[0] in [";", "."], sparql_fragment
```

Comment:
See comment above. Several occurrences in this file.

Reply:
I've replaced the `assert` statement with `raise ValueError(...)`.